### PR TITLE
feat(1654): add ListFeatureExtractor.last

### PR DESCRIPTION
**Pull request by constantine-zubrilin**
Add a "last" function and simple tests for it.

______________________________________
I confirm that I have read the [Contributor Agreements v1.0](https://github.com/robstoll/atrium/blob/main/.github/Contributor%20Agreements%20v1.0.txt), agree to be bound on them and confirm that my contribution is compliant.

### CODE REVIEW

**Comment by robstoll** [comment]
@constantine-zubrilin shall I already take a look or do you let me know once it is ready for review

**Comment by constantine-zubrilin** [comment]
Too early yet, I'll let you know, thank you

File: apis/fluent/atrium-api-fluent/src/commonMain/kotlin/ch/tutteli/atrium/api/fluent/en_GB/listExpectations.kt [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+ * Expects that the subject of `this` expectation (a [List]) is not empty and that the last element
+ * in the list holds all assertions the given [IndexWithCreator.assertionCreator] creates for it.
+ *
+ *  Use the function `last() { ... }` to create an [IndexWithCreator].
+ *
+ * @return an [Expect] for the subject of `this` expectation.
+ *
+ * @sample ch.tutteli.atrium.api.fluent.en_GB.samples.ListFeatureExtractorSamples.last
+ *
+ * @since 1.2.0
```
Comment (constantine-zubrilin) [inline apis/fluent/atrium-api-fluent/src/commonMain/kotlin/ch/tutteli/atrium/api/fluent/en_GB/listExpectations.kt:17]: Not sure that I get context well. It's a good idea to recheck comments.

**Comment by constantine-zubrilin** [comment]
@robstoll, can you please have a look?

File: apis/fluent/atrium-api-fluent/src/commonMain/kotlin/ch/tutteli/atrium/api/fluent/en_GB/listExpectations.kt [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package ch.tutteli.atrium.api.fluent.en_GB
+
+import ch.tutteli.atrium.creating.Expect
+import ch.tutteli.atrium.logic._logic
+import ch.tutteli.atrium.logic.last
+
+/**
+ * Expects that the subject of `this` expectation (a [List]) is not empty and that the last element
+ * in the list holds all assertions the given [IndexWithCreator.assertionCreator] creates for it.
```
Comment (robstoll) [inline apis/fluent/atrium-api-fluent/src/commonMain/kotlin/ch/tutteli/atrium/api/fluent/en_GB/listExpectations.kt:9]:
```suggestion
 * in the list holds all assertions the given [assertionCreator] creates for it.
```
you copied from the api-infix, here you have assertionCreator directly

File: apis/fluent/atrium-api-fluent/src/commonMain/kotlin/ch/tutteli/atrium/api/fluent/en_GB/listExpectations.kt [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+ *
+ *  Use the function `last() { ... }` to create an [IndexWithCreator].
```
Comment (robstoll) [inline apis/fluent/atrium-api-fluent/src/commonMain/kotlin/ch/tutteli/atrium/api/fluent/en_GB/listExpectations.kt:11]: Also this sentence if from the api-infix and is not necessary here as one does not need to pass an IndexWithCreator
```suggestion
```

File: apis/fluent/atrium-api-fluent/src/commonMain/kotlin/ch/tutteli/atrium/api/fluent/en_GB/listExpectations.kt [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+ *
+ * @return an [Expect] for the subject of `this` expectation.
+ *
+ * @sample ch.tutteli.atrium.api.fluent.en_GB.samples.ListFeatureExtractorSamples.last
+ *
+ * @since 1.2.0
+ */
+fun <E, T : List<E>> Expect<T>.last(assertionCreator: Expect<E>.() -> Unit): Expect<T> =
```
Comment (robstoll) [inline apis/fluent/atrium-api-fluent/src/commonMain/kotlin/ch/tutteli/atrium/api/fluent/en_GB/listExpectations.kt:19]: You forgot to add the val (first point in the issue) => found it in the correct file.
Sorry, copy&paste error on my side in the issue descriptoin, also this function should be in listFeatureExtractors.kt

File: apis/fluent/atrium-api-fluent/src/commonMain/kotlin/ch/tutteli/atrium/api/fluent/en_GB/listFeatureExtractors.kt [hunk 5]
```diff
@@ -26,3 +27,15 @@ fun <E, T : List<E>> Expect<T>.get(index: Int): Expect<E> =
 fun <E, T : List<E>> Expect<T>.get(index: Int, assertionCreator: Expect<E>.() -> Unit): Expect<T> =
     _logic.get(index).collectAndAppend(assertionCreator)
 
+/**
+ * Expects that list is not empty and returns an [Expect] for the last element in list.
+ *
+ * @return The newly created [Expect] for last element.
+ *
+ * @sample ch.tutteli.atrium.api.fluent.en_GB.samples.ListFeatureExtractorSamples.last
```
Comment (robstoll) [inline apis/fluent/atrium-api-fluent/src/commonMain/kotlin/ch/tutteli/atrium/api/fluent/en_GB/listFeatureExtractors.kt:35]:
```suggestion
 * @sample ch.tutteli.atrium.api.fluent.en_GB.samples.ListFeatureExtractorSamples.lastFeature
```

File: apis/fluent/atrium-api-fluent/src/commonTest/kotlin/ch/tutteli/atrium/api/fluent/en_GB/samples/ListFeatureExtractorSamples.kt [hunk 6]
```diff
@@ -64,4 +64,28 @@ class ListFeatureExtractorSamples {
             }
         }
     }
+
+    @Test
+    fun last() {
```
Comment (robstoll) [inline apis/fluent/atrium-api-fluent/src/commonTest/kotlin/ch/tutteli/atrium/api/fluent/en_GB/samples/ListFeatureExtractorSamples.kt:69]: split this into two please, use `last` (this fun) for explaining how the `fun last` works and `lastFeature` to explain how the `val last` works. see get and getFeature above

File: apis/infix/atrium-api-infix/src/commonMain/kotlin/ch/tutteli/atrium/api/infix/en_GB/listExpectations.kt [hunk 7]
```diff
@@ -0,0 +1,21 @@
+package ch.tutteli.atrium.api.infix.en_GB
+
+import ch.tutteli.atrium.api.infix.en_GB.creating.IndexWithCreator
+import ch.tutteli.atrium.creating.Expect
+import ch.tutteli.atrium.logic._logic
+import ch.tutteli.atrium.logic.last
+
+/**
+ * Expects that the subject of `this` expectation (a [List]) is not empty and that the last element
+ * in the list holds all assertions the given [IndexWithCreator.assertionCreator] creates for it.
```
Comment (robstoll) [inline apis/infix/atrium-api-infix/src/commonMain/kotlin/ch/tutteli/atrium/api/infix/en_GB/listExpectations.kt:10]:
```suggestion
 * in the list holds all assertions the given [assertionCreator] creates for it.
```
We also don't expect IndexWithCreator here

File: apis/infix/atrium-api-infix/src/commonMain/kotlin/ch/tutteli/atrium/api/infix/en_GB/listExpectations.kt [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+ *
+ *  Use the function `last() { ... }` to create an [IndexWithCreator].
+ *
```
Comment (robstoll) [inline apis/infix/atrium-api-infix/src/commonMain/kotlin/ch/tutteli/atrium/api/infix/en_GB/listExpectations.kt:13]:
```suggestion
```

File: apis/infix/atrium-api-infix/src/commonMain/kotlin/ch/tutteli/atrium/api/infix/en_GB/listFeatureExtractors.kt [hunk 9]
```diff
@@ -37,3 +38,15 @@ infix fun <E, T : List<E>> Expect<T>.get(index: IndexWithCreator<E>): Expect<T>
  */
 fun <E> index(index: Int, assertionCreator: Expect<E>.() -> Unit): IndexWithCreator<E> =
     IndexWithCreator(index, assertionCreator)
+
+/**
+ * Expects that list is not empty and returns an [Expect] for the last element in list.
+ *
+ * @return The newly created [Expect] for last element.
+ *
+ * @sample ch.tutteli.atrium.api.infix.en_GB.samples.ListFeatureExtractorSamples.last
```
Comment (robstoll) [inline apis/infix/atrium-api-infix/src/commonMain/kotlin/ch/tutteli/atrium/api/infix/en_GB/listFeatureExtractors.kt:47]:
```suggestion
 * @sample ch.tutteli.atrium.api.infix.en_GB.samples.ListFeatureExtractorSamples.lastFeature
```

File: translations/atrium-translations-en_GB/src/commonMain/kotlin/ch/tutteli/atrium/translations/DescriptionListLikeExpectation.kt [hunk 10]
```diff
@@ -8,5 +8,8 @@ import ch.tutteli.atrium.reporting.translating.StringBasedTranslatable
  */
 enum class DescriptionListLikeExpectation(override val value: String) : StringBasedTranslatable {
     /** @since 0.18.0 */
-    INDEX_OUT_OF_BOUNDS("❗❗ index out of bounds")
+    INDEX_OUT_OF_BOUNDS("❗❗ index out of bounds"),
+
+    /** @since 1.2.0 */
+    EMPTY("empty"),
```
Comment (robstoll) [inline translations/atrium-translations-en_GB/src/commonMain/kotlin/ch/tutteli/atrium/translations/DescriptionListLikeExpectation.kt:14]: I suggest we introduce a failure text which makes it clearer that the last element was not actually "empty" but the list was empty:
```suggestion
    WAS_EMPTY("❗❗List was empty"),
```
(for german: WAS_EMPTY("❗❗List war leer")

Other suggestions are welcome

File: misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/ListExpectationsSpec.kt [hunk 11]
```diff
@@ -82,4 +89,50 @@ abstract class ListExpectationsSpec(
...
+                    fluentNullableList.lastFun { toEqual(null) }
+                }
+            }
+        }
+    }
+
+    val emptyList = listOf<Int?>()
+    val fluentEmptyList = expect(emptyList)
+
+    val listIsEmpty = DescriptionListLikeExpectation.EMPTY.getDefault()
```
Comment (robstoll) [inline misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/ListExpectationsSpec.kt:120]:
```suggestion
    val listIsEmptyDescr = DescriptionListLikeExpectation.EMPTY.getDefault()
```

**Review by robstoll — COMMENTED** [review]
Thanks, looks already quite good, some small adjustments and we are good IMO

File: translations/atrium-translations-en_GB/src/commonMain/kotlin/ch/tutteli/atrium/translations/DescriptionListLikeExpectation.kt (hunk 10 above)
Comment (constantine-zubrilin) [reply translations/atrium-translations-en_GB/src/commonMain/kotlin/ch/tutteli/atrium/translations/DescriptionListLikeExpectation.kt:14]: I'm not sure about "was". What do you think about "IS_EMPTY"?

File: apis/fluent/atrium-api-fluent/src/commonMain/kotlin/ch/tutteli/atrium/api/fluent/en_GB/listExpectations.kt (hunk 4 above)
Comment (constantine-zubrilin) [reply apis/fluent/atrium-api-fluent/src/commonMain/kotlin/ch/tutteli/atrium/api/fluent/en_GB/listExpectations.kt]: Not sure if I got you right. As a result I removed `listExpectations.kt` totally, can you please have a look on changes?

File: apis/fluent/atrium-api-fluent/src/commonMain/kotlin/ch/tutteli/atrium/api/fluent/en_GB/listFeatureExtractors.kt [hunk 12]
```diff
@@ -26,3 +27,28 @@ fun <E, T : List<E>> Expect<T>.get(index: Int): Expect<E> =
...
+ * in the list holds all assertions the given [assertionCreator] creates for it.
+ *
+ *
+ * @return an [Expect] for the subject of `this` expectation.
+ *
+ * @sample ch.tutteli.atrium.api.fluent.en_GB.samples.ListFeatureExtractorSamples.last
+ *
+ * @since 1.2.0
+ */
+fun <E, T : List<E>> Expect<T>.last(assertionCreator: Expect<E>.() -> Unit): Expect<T> =
```
Comment (robstoll) [inline apis/fluent/atrium-api-fluent/src/commonMain/kotlin/ch/tutteli/atrium/api/fluent/en_GB/listFeatureExtractors.kt:53]: :+1: that what I meant, move last with `assertionCreator` also to this file

Comment (constantine-zubrilin) [reply apis/fluent/atrium-api-fluent/src/commonMain/kotlin/ch/tutteli/atrium/api/fluent/en_GB/listFeatureExtractors.kt:53]: Nice, thank you. So I guess it's ready to be checked by workflows.

File: apis/fluent/atrium-api-fluent/src/commonTest/kotlin/ch/tutteli/atrium/api/fluent/en_GB/samples/ListFeatureExtractorSamples.kt [hunk 13]
```diff
@@ -64,4 +64,49 @@ class ListFeatureExtractorSamples {
...
+            .last { // Subject is 3
+                toBeGreaterThan(2) // Subject is still 3
+                toBeLessThan(4) // Subject is still 3
+            }
+
+        fails {
+            expect(list).last {
+                toBeGreaterThan(3).toBeLessThan (4) // Subject is 3, fails on first expectation, second is skipped
+            }
+        }
```
Comment (robstoll) [inline apis/fluent/atrium-api-fluent/src/commonTest/kotlin/ch/tutteli/atrium/api/fluent/en_GB/samples/ListFeatureExtractorSamples.kt:104]: second isn't skipped (no fail-fast behaviour within an expectation group block) and I suggest you copy from `get` above
```suggestion
        fails {
            // all expectations are evaluated inside an expectation-group block; for more details:
            // https://github.com/robstoll/atrium#define-single-expectations-or-an-expectation-group

            expect(list).last { // subject within this expectation-group is of type Int (actually 3)
                toBeGreaterThan(3)   // fails
                toBeLessThan (4)     // still evaluated, even though  `toBeGreaterThan` already fails,
                //                      use `.last.` if you want a fail fast behaviour
            } // subject here is back type List<Int>
        }
```

File: apis/infix/atrium-api-infix/src/commonTest/kotlin/ch/tutteli/atrium/api/infix/en_GB/samples/ListFeatureExtractorSamples.kt [hunk 14]
```diff
@@ -58,4 +57,48 @@ class ListFeatureExtractorSamples {
...
+            }
+            .last {
+                it toBeGreaterThan(2)
+                it toBeLessThan(4) // Subject is 3 and passes all expectations
+            }
+
+        fails {
+            expect(list).last {
+                it toBeGreaterThan (3)
+                it toBeLessThan (4) // Subject is 3, fails on first expectation, second is skipped
```
Comment (robstoll) [inline apis/infix/atrium-api-infix/src/commonTest/kotlin/ch/tutteli/atrium/api/infix/en_GB/samples/ListFeatureExtractorSamples.kt:94]: same same as in api-fluent

File: translations/atrium-translations-de_CH/src/commonMain/kotlin/ch/tutteli/atrium/translations/DescriptionListLikeExpectation.kt [hunk 15]
```diff
@@ -8,5 +8,8 @@ import ch.tutteli.atrium.reporting.translating.StringBasedTranslatable
  */
 enum class DescriptionListLikeExpectation(override val value: String) : StringBasedTranslatable {
     /** @since 0.18.0 */
-    INDEX_OUT_OF_BOUNDS("❗❗ Index ausserhalb der Grenzen (index out of bounds)")
+    INDEX_OUT_OF_BOUNDS("❗❗ Index ausserhalb der Grenzen (index out of bounds)"),
+
+    /** @since 1.2.0 */
+    IS_EMPTY("❗❗Liste ist leer (list is empty)"),
```
Comment (robstoll) [inline translations/atrium-translations-de_CH/src/commonMain/kotlin/ch/tutteli/atrium/translations/DescriptionListLikeExpectation.kt:14]:
```suggestion
    IS_EMPTY("❗❗ Liste ist leer (list is empty)"),
```

File: translations/atrium-translations-en_GB/src/commonMain/kotlin/ch/tutteli/atrium/translations/DescriptionListLikeExpectation.kt [hunk 16]
```diff
@@ -8,5 +8,8 @@ import ch.tutteli.atrium.reporting.translating.StringBasedTranslatable
  */
 enum class DescriptionListLikeExpectation(override val value: String) : StringBasedTranslatable {
     /** @since 0.18.0 */
-    INDEX_OUT_OF_BOUNDS("❗❗ index out of bounds")
+    INDEX_OUT_OF_BOUNDS("❗❗ Index out of bounds"),
+
+    /** @since 1.2.0 */
+    IS_EMPTY("❗❗List is empty"),
```
Comment (robstoll) [inline translations/atrium-translations-en_GB/src/commonMain/kotlin/ch/tutteli/atrium/translations/DescriptionListLikeExpectation.kt:14]:
```suggestion
    IS_EMPTY("❗❗ List is empty"),
```

**Review by robstoll — CHANGES_REQUESTED** [review]

File: apis/fluent/atrium-api-fluent/src/commonMain/kotlin/ch/tutteli/atrium/api/fluent/en_GB/listExpectations.kt (hunk 4 above)
Comment (robstoll) [reply apis/fluent/atrium-api-fluent/src/commonMain/kotlin/ch/tutteli/atrium/api/fluent/en_GB/listExpectations.kt]: @constantine-zubrilin Almost there, a few little things and we are good

File: apis/infix/atrium-api-infix/src/commonTest/kotlin/ch/tutteli/atrium/api/infix/en_GB/samples/ListFeatureExtractorSamples.kt (hunk 14 above)
Comment (constantine-zubrilin) [reply apis/infix/atrium-api-infix/src/commonTest/kotlin/ch/tutteli/atrium/api/infix/en_GB/samples/ListFeatureExtractorSamples.kt]: Fixed

**Review by robstoll — APPROVED** [review]

**Comment by robstoll** [comment]
@constantine-zubrilin thanks for your first contribution to Atrium 🙂👍

**Pull request merged**